python/glow/wgr/linear_model/logistic_model.py

```diff
@@ -0,0 +1,245 @@
+import pyspark.sql.functions as f
+from pyspark.sql.window import Window
+from typeguard import typechecked
+from typing import Any, Dict, List
+from glow.logging import record_hls_event
+
+
+@typechecked
+class LogisticRegression:
+    """
+    The LogisticRegression class is used to fit logistic regression models against one or labels optimized over a
+    provided list of ridge alpha parameters. The optimal ridge alpha value is chosen for each label by minimizing the
+    average out of fold log_loss scores.
+    """
+    def __init__(self, alphas: NDArray[(Any, ), Float] = np.array([])) -> None:
+        """
+        LogisticRegression is initialized with a list of alpha values.
+
+        Args:
+            alphas : array_like of alpha values used in the ridge regression (optional).
+        """
+        if not (alphas >= 0).all():
+            raise Exception('Alpha values must all be non-negative.')
+        self.alphas = create_alpha_dict(alphas)
+
+    def fit(self, blockdf: DataFrame, labeldf: pd.DataFrame, sample_blocks: Dict[str, List[str]],
+            covdf: pd.DataFrame) -> (DataFrame, DataFrame):
+        """
+        Fits a logistic regression model, represented by a Spark DataFrame containing coefficients for each of the ridge
+        alpha parameters, for each block in the starting matrix, for each label in the target labels, as well as a
+        Spark DataFrame containing the optimal ridge alpha value for each label.
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble (optional).
+
+        Returns:
+            Two Spark DataFrames, one containing the model resulting from the fitting routine and one containing the
+            results of the cross validation procedure.
+        """
+        map_key_pattern = ['sample_block', 'label', 'alpha_name']
+        reduce_key_pattern = ['header_block', 'header', 'label', 'alpha_name']
+        model_key_pattern = ['sample_block', 'label']
+        p0_dict = {k: v for k, v in zip(labeldf.columns, labeldf.sum(axis=0) / labeldf.shape[0])}
+
+        if not self.alphas:
+            self.alphas = generate_alphas(blockdf)
+
+        map_udf = pandas_udf(
+            lambda key, pdf: map_irls_eqn(key, map_key_pattern, pdf, labeldf, sample_blocks, covdf,
+                                          p0_dict, self.alphas), irls_eqn_struct,
+            PandasUDFType.GROUPED_MAP)
+
+        reduce_udf = pandas_udf(lambda key, pdf: reduce_irls_eqn(key, reduce_key_pattern, pdf),
+                                irls_eqn_struct, PandasUDFType.GROUPED_MAP)
+
+        model_udf = pandas_udf(
```

Comment:
I think I understand what you mean.  The issue is that we ultimately want the coefficient estimates for a particular header from all alphas collected on a single row, so if we grouped by alphas as well we would have to do some extra work to get the coefficients back together again.  It's probably a better design though and would lead to better parallelism like you said, I'll take a pass at trying it that way.

Reply:
I updated the logic to group by `alpha_name` as well, followed by some spark transformations to produce the correct final format.  Take a look and see if it makes more sense.

Reply:
Thanks, I like this. In addition to improving parallelism, it makes the UDFs, where most of the work happens, a lot easier to understand.